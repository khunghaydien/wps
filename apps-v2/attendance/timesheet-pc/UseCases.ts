import UseCaseCollection from '@attendance/application/UseCaseCollection';
import { IUseCase as ISubmitFixDailyRequestAndSaveDailyRecordUseCase } from '@attendance/domain/combinedUseCases/ISaveDailyRecordAndSubmitFixDailyRequestUseCase';
import { IUseCase as ISubmitFixDailyRequestWithClockOutUseCase } from '@attendance/domain/combinedUseCases/ISubmitFixDailyRequestWithClockOutUseCase';
import { IUseCase as IDailyRecordFetchRestTimeReasonsUseCase } from '@attendance/domain/useCases/dailyRecord/IFetchRestTimeReasonsUseCase';
import { IUseCase as IDailyRecordSaveFieldsUseCase } from '@attendance/domain/useCases/dailyRecord/ISaveFieldsUseCase';
import { IUseCase as IDailyRecordSaveUseCase } from '@attendance/domain/useCases/dailyRecord/ISaveUseCase';
import { IUseCase as IFixDailyRequestCancelApprovalUseCase } from '@attendance/domain/useCases/fixDailyRequest/ICancelApprovalUseCase';
import { IUseCase as IFixDailyRequestCancelSubmittedUseCase } from '@attendance/domain/useCases/fixDailyRequest/ICancelSubmittedUseCase';
import { IUseCase as IFixDailyRequestSubmitUseCase } from '@attendance/domain/useCases/fixDailyRequest/ISubmitUseCase';
import { IUseCase as ILegalAgreementFetchOvertimeUseCase } from '@attendance/domain/useCases/legalAgreement/IFetchOvertimeUseCase';
import { IUseCase as ILegalAgreementRequestApprovalRequestUseCase } from '@attendance/domain/useCases/legalAgreementRequest/ICancelApprovalUseCase';
import { IUseCase as ILegalAgreementRequestCancelRequestUseCase } from '@attendance/domain/useCases/legalAgreementRequest/ICancelRequestUseCase';
import { IUseCase as ILegalAgreementRequestFetchListUseCase } from '@attendance/domain/useCases/legalAgreementRequest/IFetchListUseCase';
import { IUseCase as ILegalAgreementRequestReapplyUseCase } from '@attendance/domain/useCases/legalAgreementRequest/IReapplyUseCase';
import { IUseCase as ILegalAgreementRequestRemoveUseCase } from '@attendance/domain/useCases/legalAgreementRequest/IRemoveUseCase';
import { IUseCase as ILegalAgreementRequestSubmitUseCase } from '@attendance/domain/useCases/legalAgreementRequest/ISubmitUseCase';
import { IUseCase as IObjectivelyEventLogCreateUseCase } from '@attendance/domain/useCases/objectivelyEventLog/ICreateUseCase';
import { IUseCase as IObjectivelyEventLogFetchDailyDeviationReasonsUseCase } from '@attendance/domain/useCases/objectivelyEventLog/IFetchDailyDeviationReasonsUseCase';
import { IUseCase as IObjectivelyEventLogFetchDailyUseCase } from '@attendance/domain/useCases/objectivelyEventLog/IFetchDailyUseCase';
import { IUseCase as IObjectivelyEventLogFetchUseCase } from '@attendance/domain/useCases/objectivelyEventLog/IFetchUseCase';
import { IUseCase as IObjectivelyEventLogRemoveUseCase } from '@attendance/domain/useCases/objectivelyEventLog/IRemoveUseCase';
import { IUseCase as IObjectivelyEventLogSaveDailyDeviationReasonUseCase } from '@attendance/domain/useCases/objectivelyEventLog/ISaveDailyDeviationReasonUseCase';
import { IUseCase as IObjectivelyEventLogSetToBeAppliedToDailyUseCase } from '@attendance/domain/useCases/objectivelyEventLog/ISetToBeAppliedToDailyUseCase';
import { IUseCase as IStampTimePostUseCase } from '@attendance/domain/useCases/stampTime/IPostUseCase';
import { IUseCase as IDisplayFieldLayoutFetchTableUseCase } from '@attendance/domain/useCases/timesheet/IFetchTableUseCase';
import { IUseCase as ITimesheetFetchUseCase } from '@attendance/domain/useCases/timesheet/IFetchUseCase';

export type UseCases = {
  fetchTimesheet: ITimesheetFetchUseCase;
  reloadTimesheetOnly: ITimesheetFetchUseCase;
  fetchObjectivelyEventLogs: IObjectivelyEventLogFetchUseCase;
  fetchDailyObjectivelyEventLogDeviationReasons: IObjectivelyEventLogFetchDailyDeviationReasonsUseCase;
  setToBeAppliedToDailyObjectivelyEventLog: IObjectivelyEventLogSetToBeAppliedToDailyUseCase;
  reloadDailyObjectivelyEventLogs: IObjectivelyEventLogFetchDailyUseCase;
  reloadOneDailyObjectivelyEventLog: IObjectivelyEventLogFetchDailyUseCase;
  createObjectivelyEventLog: IObjectivelyEventLogCreateUseCase;
  removeObjectivelyEventLog: IObjectivelyEventLogRemoveUseCase;
  submitFixDailyRequest: IFixDailyRequestSubmitUseCase;
  submitFixDailyRequestWithClockOut: ISubmitFixDailyRequestWithClockOutUseCase;
  submitFixDailyRequestAndSaveDailyRecord: ISubmitFixDailyRequestAndSaveDailyRecordUseCase;
  cancelSubmittedFixDailyRequest: IFixDailyRequestCancelSubmittedUseCase;
  cancelApprovalFixDailyRequest: IFixDailyRequestCancelApprovalUseCase;
  saveDailyRecord: IDailyRecordSaveUseCase;
  saveDailyObjectivelyEventLogDeviationReason: IObjectivelyEventLogSaveDailyDeviationReasonUseCase;
  stampTime: IStampTimePostUseCase;
  fetchRestTimeReasons: IDailyRecordFetchRestTimeReasonsUseCase;
  fetchListLegalAgreementRequest: ILegalAgreementRequestFetchListUseCase;
  fetchOvertimeLegalAgreement: ILegalAgreementFetchOvertimeUseCase;
  submitLegalAgreementRequest: ILegalAgreementRequestSubmitUseCase;
  cancelRequestLegalAgreementRequest: ILegalAgreementRequestCancelRequestUseCase;
  cancelApprovalLegalAgreementRequest: ILegalAgreementRequestApprovalRequestUseCase;
  removeLegalAgreementRequest: ILegalAgreementRequestRemoveUseCase;
  reapplyLegalAgreementRequest: ILegalAgreementRequestReapplyUseCase;
  fetchDailyFieldLayoutTable: IDisplayFieldLayoutFetchTableUseCase;
  saveFields: IDailyRecordSaveFieldsUseCase;
};

const service = UseCaseCollection<UseCases>("Timesheet-PC's UseCase Service");

export default service;
