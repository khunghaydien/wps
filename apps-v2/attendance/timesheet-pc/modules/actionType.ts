export default 'TIMESHEET-PC' as const;
