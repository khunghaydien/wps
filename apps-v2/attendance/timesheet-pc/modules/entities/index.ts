import { combineReducers } from 'redux';

import dailyAllowance from './dailyAllowance';
import dailyDeviatedReason from './dailyDeviatedReason';
import dailyTimeTrack from './dailyTimeTrack';
import objectivelyEventLog from './objectivelyEventLog';
import restTimeReasons from './restTimeReasons';
import stampWidget from './stampWidget';
import timesheet from './timesheet';
import timeTrackAlert from './timeTrackAlert';

const reducers = {
  dailyTimeTrack,
  stampWidget,
  timesheet,
  timeTrackAlert,
  dailyAllowance,
  objectivelyEventLog,
  restTimeReasons,
  dailyDeviatedReason,
};

const rootReducer = combineReducers(reducers);
export type State = ReturnType<typeof rootReducer>;

export default rootReducer;
