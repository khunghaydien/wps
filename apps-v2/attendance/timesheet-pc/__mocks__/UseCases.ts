import { UseCases } from '../UseCases';

export const methods: UseCases = {
  fetchTimesheet: jest.fn().mockResolvedValue({
    timesheet: 'timesheet',
  }),
  reloadTimesheetOnly: jest.fn().mockResolvedValue({
    timesheet: 'timesheet',
  }),
  fetchObjectivelyEventLogs: jest.fn().mockResolvedValue({
    objectivelyEventLogs: 'objectivelyEventLogs',
  }),
  setToBeAppliedToDailyObjectivelyEventLog: jest
    .fn()
    .mockResolvedValue('result'),
  reloadDailyObjectivelyEventLogs: jest.fn().mockResolvedValue({
    dailyObjectivelyEventLogs: ['dailyObjectivelyEventLog'],
  }),
  reloadOneDailyObjectivelyEventLog: jest.fn().mockResolvedValue({
    dailyObjectivelyEventLogs: ['dailyObjectivelyEventLog'],
  }),
  createObjectivelyEventLog: jest.fn().mockResolvedValue('result'),
  removeObjectivelyEventLog: jest.fn().mockResolvedValue('result'),
  stampTime: jest.fn().mockResolvedValue('result'),
  saveDailyRecord: jest.fn().mockResolvedValue('result'),
  saveDailyObjectivelyEventLogDeviationReason: jest
    .fn()
    .mockResolvedValue('result'),
  submitFixDailyRequest: jest.fn().mockResolvedValue('result'),
  submitFixDailyRequestWithClockOut: jest.fn().mockRejectedValue('result'),
  submitFixDailyRequestAndSaveDailyRecord: jest
    .fn()
    .mockRejectedValue('result'),
  cancelSubmittedFixDailyRequest: jest.fn().mockResolvedValue('result'),
  cancelApprovalFixDailyRequest: jest.fn().mockResolvedValue('result'),
  fetchRestTimeReasons: jest.fn().mockResolvedValue('result'),
  fetchDailyObjectivelyEventLogDeviationReasons: jest
    .fn()
    .mockResolvedValue('result'),
  fetchListLegalAgreementRequest: jest.fn().mockResolvedValue('result'),
  fetchOvertimeLegalAgreement: jest.fn().mockResolvedValue('result'),
  submitLegalAgreementRequest: jest.fn().mockResolvedValue('result'),
  reapplyLegalAgreementRequest: jest.fn().mockResolvedValue('result'),
  cancelRequestLegalAgreementRequest: jest.fn().mockResolvedValue('result'),
  cancelApprovalLegalAgreementRequest: jest.fn().mockResolvedValue('result'),
  removeLegalAgreementRequest: jest.fn().mockResolvedValue('result'),
  fetchDailyFieldLayoutTable: jest.fn().mockResolvedValue('result'),
  saveFields: jest.fn().mockResolvedValue('result'),
};

export default (): UseCases => methods;
