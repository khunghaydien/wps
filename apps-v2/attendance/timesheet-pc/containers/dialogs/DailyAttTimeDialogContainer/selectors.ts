import { State } from '../../../modules';

export const localLoading = (state: State) =>
  state.ui.loadingDailyObjectivelyEventLog;
