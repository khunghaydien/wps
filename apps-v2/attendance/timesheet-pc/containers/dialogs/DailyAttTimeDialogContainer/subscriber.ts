import { Store } from 'redux';

import { clear as clearRestReasons } from '../../../modules/entities/restTimeReasons';
import { close } from '../../../modules/ui/dailyAttTimeDialog';
import { actions as editingDailyAttTimeActions } from '../../../modules/ui/editingDailyAttTime';

import Events from '../../../events';
import LocalEvents from './events';

export default (store: Store): (() => void) => {
  const { dispatch } = store;

  let updated = false;
  const exit = (result: boolean) => {
    updated = true;
    if (result) {
      dispatch(close());
    }
  };
  const unsubscribers = [];
  unsubscribers.push(
    // 申請ボタンを押したとき
    LocalEvents.submittedRequest.subscribe(exit),
    // 保存ボタンを押したとき
    LocalEvents.saved.subscribe(exit),
    // 承認取り消しをしたとき
    LocalEvents.canceledApprovalRequest.subscribe(
      Events.updatedDailyRecord.publish
    ),
    // 申請取り消しをしたとき
    LocalEvents.canceledSubmittedRequest.subscribe(
      Events.updatedDailyRecord.publish
    )
  );

  return () => {
    unsubscribers.forEach((f) => f());
    dispatch(editingDailyAttTimeActions.unset());
    dispatch(clearRestReasons());
    if (updated) {
      Events.updatedDailyRecord.publish();
    }
  };
};
