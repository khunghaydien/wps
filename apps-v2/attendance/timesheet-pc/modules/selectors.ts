import isNil from 'lodash/isNil';
import { createSelector } from 'reselect';

import {
  isPermissionSatisfied,
  Permission,
} from '../../../domain/models/access-control/Permission';
import AttRecord from '../models/AttRecord';
import DailyRequestConditions from '../models/DailyRequestConditions';
import * as AttDailyRequest from '@attendance/domain/models/AttDailyRequest';
import { DailyRequestNameMap } from '@attendance/domain/models/AttDailyRequestType';
import * as AttDailyRequestType from '@attendance/domain/models/AttDailyRequestType';
import {
  buildDailyActualWorkingPeriodListFromAttRecord,
  DailyActualWorkingTimePeriod,
} from '@attendance/domain/models/DailyActualWorkingTimePeriod';
import { AttSummary } from '@attendance/domain/models/DeprecatedAttSummary';
import * as TimeRange from '@attendance/domain/models/TimeRange';

import { State as TimesheetState } from './entities/timesheet';
import { State } from './index';

const hasPermissionForEdit = ({
  isByDelegate,
  userPermission,
}: {
  isByDelegate: boolean;
  userPermission: Permission;
}): boolean =>
  isPermissionSatisfied({
    isByDelegate,
    userPermission,
    allowIfByEmployee: true,
    requireIfByDelegate: ['editAttTimeSheetByDelegate'],
  });

export const isLoading = (state: State): boolean =>
  !!state.common.app.loadingDepth;

const selectIsSummaryLocked = (state: State): boolean =>
  !isNil(state.entities.timesheet.attSummary) &&
  state.entities.timesheet.attSummary.isLocked;

export const isByDelegate = (state: State): boolean =>
  state.common.proxyEmployeeInfo.isProxyMode;

export const employeeId: (state: State) => string = createSelector(
  (state: State) => state.common.userSetting.employeeId,
  (state: State) => state.common.proxyEmployeeInfo.id,
  isByDelegate,
  (employeeId, proxyEmployeeId, isByDelegate) =>
    isByDelegate ? proxyEmployeeId : employeeId
);

const selectUserPermission = (state: State): Permission =>
  state.common.accessControl.permission;

export const isTimesheetReadOnly: (arg0: State) => boolean = createSelector(
  selectIsSummaryLocked,
  isByDelegate,
  selectUserPermission,
  (isSummaryLocked, isByDelegate, userPermission) =>
    isSummaryLocked || !hasPermissionForEdit({ isByDelegate, userPermission })
);

const selectAttSummary = (state: State): null | AttSummary =>
  (state.entities.timesheet as TimesheetState).attSummary;
const selectAttRecordList = (state: State) =>
  (state.entities.timesheet as TimesheetState).attRecordList;
const selectAttDailyRequestMap = (state: State) =>
  (state.entities.timesheet as TimesheetState).attDailyRequestMap;
const selectAttDailyRequestTypeMap = (state: State) =>
  (state.entities.timesheet as TimesheetState).attDailyRequestTypeMap;

// 実労働時間をモデルのインスタンスとして構築して、日付をキーにしたマップとして返却する
// TODO: 処理タイミングを再考したい。
// - store格納時：state上では正規化して配置されるべき？
// - AttRecordのコンストラクタ：データの更新と必ず同期するが、無駄な処理が増える懸念がある
// - AttRecordのcreateFromParamメソッド：前項に比べて処理負荷は軽減されるが、同期漏れが発生しうる
export const buildDailyActualWorkingPeriodListMap: (arg0: State) => {
  [key: string]: DailyActualWorkingTimePeriod[];
} = createSelector(
  // $FlowFixMe v0.85
  selectAttRecordList,
  (attRecordList: AttRecord[]) => {
    if (!attRecordList) {
      return {};
    }

    return attRecordList.reduce((hash, attRecord) => {
      hash[attRecord.recordDate] =
        buildDailyActualWorkingPeriodListFromAttRecord(attRecord);
      return hash;
    }, {});
  }
);

// 予定されている勤務時刻をモデルのインスタンスとして構築して、日付をキーにしたマップとして返却する
// TODO: 処理タイミングを再考したい。※詳細は同上
export const buildDailyRequestedWorkingHoursMap: (arg0: State) => {
  [key: string]: TimeRange.TimeRange;
} = createSelector(
  selectAttDailyRequestMap,
  (attDailyRequestMap: Record<string, AttDailyRequest.AttDailyRequest>) => {
    const dailyRequestedWorkingHoursMap = Object.keys(attDailyRequestMap || {})
      .map((requestId) => attDailyRequestMap[requestId])
      .filter((request) =>
        AttDailyRequest.isEffectual(
          request,
          AttDailyRequestType.CODE.HolidayWork
        )
      )
      .reduce((hash, request) => {
        hash[request.startDate] = TimeRange.convert(request);
        return hash;
      }, {});

    return dailyRequestedWorkingHoursMap;
  }
);

// 日ごとの申請の状態をモデルのインスタンスとして構築して、日付をキーにしたマップとして返却する
export const buildDailyRequestConditionMap: (arg0: State) => {
  [key: string]: DailyRequestConditions;
} = createSelector(
  selectAttSummary,
  selectAttRecordList,
  selectAttDailyRequestMap,
  selectAttDailyRequestTypeMap,
  isByDelegate,
  selectUserPermission,
  (
    attSummary: AttSummary,
    attRecordList: AttRecord[],
    attDailyRequestMap: { [key: string]: AttDailyRequest.AttDailyRequest },
    attDailyRequestTypeMap: DailyRequestNameMap,
    isByDelegate: boolean,
    userPermission: Permission
  ) => {
    if (!attRecordList) {
      return {};
    }

    return attRecordList.reduce((hash, attRecord) => {
      hash[attRecord.recordDate] = DailyRequestConditions.createFromParams(
        attRecord,
        attDailyRequestMap,
        attDailyRequestTypeMap,
        {
          isSummaryLocked: isNil(attSummary) ? false : attSummary.isLocked,
          isByDelegate,
          userPermission,
        }
      );
      return hash;
    }, {});
  }
);

// @ts-ignore Remove this line after TS migration finished
const selectTargetDate = (state: State) => state.ui.dailyRequest.targetDate;

export const selectSelectedRequestConditions: (
  arg0: State
) => DailyRequestConditions = createSelector(
  buildDailyRequestConditionMap,
  selectTargetDate,
  (
    dailyRequestConditionsMap: {
      [key: string]: DailyRequestConditions;
    },
    targetDate: string
  ) => dailyRequestConditionsMap[targetDate]
);

const editingAttRecordId = (state: State): string =>
  state.ui.dailyAttTimeDialog;

export const editingAttRecord = createSelector(
  editingAttRecordId,
  selectAttRecordList,
  (id, records) => records?.find((record) => record.id === id)
);

export const editingDailyObjectivelyEventLog = createSelector(
  editingAttRecordId,
  (state: State) => state.entities.timesheet.dailyObjectivelyEventLogs,
  (recordId, records) => records?.find((record) => record.recordId === recordId)
);

const isDailyRecordLocked = (recordId: string) =>
  createSelector(
    selectAttRecordList,
    (records) => records?.find((record) => record.id === recordId)?.isLocked
  );

export const isDailyRecordReadOnly = (
  recordId: string
): ((state: State) => boolean) =>
  createSelector(
    isTimesheetReadOnly,
    isDailyRecordLocked(recordId),
    (isSummaryLocked, isDailyRecordLocked) =>
      isSummaryLocked || isDailyRecordLocked
  );
