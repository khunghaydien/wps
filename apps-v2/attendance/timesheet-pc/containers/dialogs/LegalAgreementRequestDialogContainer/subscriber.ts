import { Store } from 'redux';

import { State } from '@attendance/timesheet-pc/modules';
import * as selectors from '@attendance/timesheet-pc/modules/selectors';

import actions from './actions';

import { IOutputData as IFetchListUseCaseOutputData } from '@attendance/domain/useCases/legalAgreementRequest/IFetchListUseCase';
import Events from '@attendance/timesheet-pc/events';
import UseCases from '@attendance/timesheet-pc/UseCases';

export default (store: Store): (() => void) => {
  let updated = false;
  const update = () => {
    updated = true;
  };

  const reload = async () => {
    const state = store.getState() as State;
    const targetDate = state.entities.timesheet.ownerInfos?.at(-1)?.endDate;
    const employeeId = selectors.employeeId(state) || null;
    UseCases().fetchListLegalAgreementRequest({
      employeeId,
      targetDate,
    });
    UseCases().fetchOvertimeLegalAgreement({
      employeeId,
      targetDate,
    });
    update();
  };

  const setDetail = async (result: IFetchListUseCaseOutputData) => {
    const state = store.getState() as State;
    const targetId = state.ui.legalAgreementRequest.editing.id;
    const selectedRequest = result?.requestList?.requests?.find(
      (item) => item?.id === targetId
    );
    if (selectedRequest) {
      actions(store.dispatch).showRequestDetailPane(selectedRequest);
    }
  };

  const unsubscribers = [];
  unsubscribers.push(
    UseCases().submitLegalAgreementRequest.subscribe(update),
    UseCases().reapplyLegalAgreementRequest.subscribe(update),
    UseCases().cancelRequestLegalAgreementRequest.subscribe(reload),
    UseCases().cancelApprovalLegalAgreementRequest.subscribe(reload),
    UseCases().removeLegalAgreementRequest.subscribe((result) => {
      if (result) {
        reload();
      }
    }),
    UseCases().fetchListLegalAgreementRequest.subscribe(setDetail)
  );

  return () => {
    unsubscribers.forEach((f) => f());
    // 開いて何もせずに閉じたときに勤務表読み込みが走るのを抑止している
    if (updated) {
      Events.updatedDailyRecord.publish();
    }
  };
};
