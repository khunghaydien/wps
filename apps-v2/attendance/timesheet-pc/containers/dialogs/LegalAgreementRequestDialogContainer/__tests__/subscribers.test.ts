import configureMockStore from 'redux-mock-store';

import * as selectors from '@attendance/timesheet-pc/modules/selectors';

import subscriber from '../subscriber';
import EventEmitter from '@attendance/libraries/Event/emitter';
import Events from '@attendance/timesheet-pc/events';
import UseCases, {
  // @ts-ignore
  mocked as UseCaseMockedMethods,
  UseCases as UseCaseMethods,
} from '@attendance/timesheet-pc/UseCases';

const createStore = configureMockStore();

jest.mock('@attendance/timesheet-pc/modules/selectors', () => ({
  __esModule: true,
  employeeId: jest.fn(),
}));

jest.mock('@attendance/timesheet-pc/UseCases', () => {
  const act = jest.requireActual('../../../../UseCases');
  const mocked = jest.requireActual('../../../../__mocks__/UseCases').methods;
  act.default.register(mocked);
  return {
    __esModule: true,
    default: act.default,
    mocked,
  };
});

beforeEach(() => {
  jest.clearAllMocks();
});

it.each([
  'cancelRequestLegalAgreementRequest',
  'cancelApprovalLegalAgreementRequest',
] as (keyof Pick<
  UseCaseMethods,
  'cancelRequestLegalAgreementRequest' | 'cancelApprovalLegalAgreementRequest'
>)[])('should be call reload when finished %s', (functionName) => {
  // Arrange
  const publish = jest.spyOn(EventEmitter, 'publish');
  (selectors.employeeId as jest.Mock).mockReturnValue('employeeId');
  const store = createStore({
    ui: {
      legalAgreementRequest: {
        editing: {
          id: 'xxx',
        },
      },
    },
    entities: {
      timesheet: {
        ownerInfos: [
          {
            endDate: '2022-02-22',
          },
        ],
      },
    },
  });
  const unsubscriber = subscriber(store);

  // Act
  EventEmitter.publish(UseCases()[functionName].eventName);
  unsubscriber();

  // Assert
  expect(
    UseCaseMockedMethods.fetchListLegalAgreementRequest
  ).toHaveBeenCalledTimes(1);
  expect(
    UseCaseMockedMethods.fetchListLegalAgreementRequest
  ).toHaveBeenCalledWith({
    employeeId: 'employeeId',
    targetDate: '2022-02-22',
  });
  expect(
    UseCaseMockedMethods.fetchOvertimeLegalAgreement
  ).toHaveBeenCalledTimes(1);
  expect(UseCaseMockedMethods.fetchOvertimeLegalAgreement).toHaveBeenCalledWith(
    {
      employeeId: 'employeeId',
      targetDate: '2022-02-22',
    }
  );

  expect(publish).toHaveBeenCalledTimes(2);
  expect(publish).toHaveBeenNthCalledWith(
    2,
    Events.updatedDailyRecord.eventName,
    undefined
  );
});

describe.each(['removeLegalAgreementRequest'] as (keyof Pick<
  UseCaseMethods,
  'removeLegalAgreementRequest'
>)[])('%s', (functionName) => {
  it('should not call exit if result is false', () => {
    // Arrange
    const publish = jest.spyOn(EventEmitter, 'publish');
    (selectors.employeeId as jest.Mock).mockReturnValue('employeeId');
    const store = createStore({
      ui: {
        legalAgreementRequest: {
          editing: {
            id: 'xxx',
          },
        },
      },
      entities: {
        timesheet: {
          ownerInfos: [
            {
              endDate: '2022-02-22',
            },
          ],
        },
      },
    });
    const unsubscriber = subscriber(store);

    // Act
    EventEmitter.publish(UseCases()[functionName].eventName);
    unsubscriber();

    // Assert
    expect(
      UseCaseMockedMethods.fetchListLegalAgreementRequest
    ).toHaveBeenCalledTimes(0);
    expect(
      UseCaseMockedMethods.fetchOvertimeLegalAgreement
    ).toHaveBeenCalledTimes(0);

    expect(publish).toHaveBeenCalledTimes(1);
  });

  it('should not call exit if result is true', () => {
    // Arrange
    const publish = jest.spyOn(EventEmitter, 'publish');
    (selectors.employeeId as jest.Mock).mockReturnValue('employeeId');
    const store = createStore({
      ui: {
        legalAgreementRequest: {
          editing: {
            id: 'xxx',
          },
        },
      },
      entities: {
        timesheet: {
          ownerInfos: [
            {
              endDate: '2022-02-22',
            },
          ],
        },
      },
    });
    const unsubscriber = subscriber(store);

    // Act
    EventEmitter.publish(UseCases()[functionName].eventName, true);
    unsubscriber();

    // Assert
    expect(
      UseCaseMockedMethods.fetchListLegalAgreementRequest
    ).toHaveBeenCalledTimes(1);
    expect(
      UseCaseMockedMethods.fetchListLegalAgreementRequest
    ).toHaveBeenCalledWith({
      employeeId: 'employeeId',
      targetDate: '2022-02-22',
    });
    expect(
      UseCaseMockedMethods.fetchOvertimeLegalAgreement
    ).toHaveBeenCalledTimes(1);
    expect(
      UseCaseMockedMethods.fetchOvertimeLegalAgreement
    ).toHaveBeenCalledWith({
      employeeId: 'employeeId',
      targetDate: '2022-02-22',
    });

    expect(publish).toHaveBeenCalledTimes(2);
    expect(publish).toHaveBeenNthCalledWith(
      2,
      Events.updatedDailyRecord.eventName,
      undefined
    );
  });
});

describe.each([
  'submitLegalAgreementRequest',
  'reapplyLegalAgreementRequest',
] as (keyof Pick<
  UseCaseMethods,
  'submitLegalAgreementRequest' | 'reapplyLegalAgreementRequest'
>)[])('%s', (functionName) => {
  it('should be call exit when finished %s', () => {
    const store = createStore({
      ui: {
        legalAgreementRequest: {
          editing: {
            id: 'xxx',
          },
        },
      },
      entities: {
        timesheet: {
          ownerInfos: [
            {
              endDate: '2022-02-22',
            },
          ],
        },
      },
    });
    const publish = jest.spyOn(EventEmitter, 'publish');
    const unsubscriber = subscriber(store);
    EventEmitter.publish(UseCases()[functionName].eventName);

    // Act
    unsubscriber();

    // Assert
    expect(publish).toHaveBeenCalledTimes(2);
    expect(publish).toHaveBeenNthCalledWith(
      2,
      Events.updatedDailyRecord.eventName,
      undefined
    );
  });
});

describe.each(['fetchListLegalAgreementRequest'] as (keyof Pick<
  UseCaseMethods,
  'fetchListLegalAgreementRequest'
>)[])('%s', (functionName) => {
  it('should be call select when had targetId', async () => {
    // Arrange
    const store = createStore({
      ui: {
        legalAgreementRequest: {
          editing: {
            id: 'abc',
          },
        },
      },
      entities: {
        timesheet: {
          ownerInfos: [
            {
              endDate: '2022-02-22',
            },
          ],
        },
      },
    });
    const unsubscriber = subscriber(store);

    // Act
    await EventEmitter.publish(UseCases()[functionName].eventName, {
      requestList: {
        requests: [
          {
            id: 'abc',
          },
        ],
      },
    });
    unsubscriber();

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });

  it('should be not call select when had not targetId', async () => {
    // Arrange
    const store = createStore({
      ui: {
        legalAgreementRequest: {
          editing: {},
        },
      },
      entities: {
        timesheet: {
          ownerInfos: [
            {
              endDate: '2022-02-22',
            },
          ],
        },
      },
    });
    const unsubscriber = subscriber(store);

    // Act
    await EventEmitter.publish(UseCases()[functionName].eventName, {
      requestList: {
        requests: [
          {
            id: 'abc',
          },
        ],
      },
    });
    unsubscriber();

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });
});
