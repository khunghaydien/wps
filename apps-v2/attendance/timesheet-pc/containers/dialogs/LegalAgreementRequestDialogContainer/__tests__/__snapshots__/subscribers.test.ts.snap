// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchListLegalAgreementRequest should be call select when had targetId 1`] = `
Array [
  Object {
    "payload": Object {
      "request": Object {
        "id": "abc",
      },
    },
    "type": "TIMESHEET-PC/UI/LEGAL-AGREEMENT-REQUEST/REQUESTS/MONTHLY-REQUEST/INIT",
  },
  Object {
    "payload": Object {
      "isEditing": false,
      "request": Object {
        "id": "abc",
      },
    },
    "type": "TIMESHEET-PC/UI/LEGAL-AGREEMENT-REQUEST/EDITING/INIT",
  },
]
`;

exports[`fetchListLegalAgreementRequest should be not call select when had not targetId 1`] = `Array []`;
